guides/v2.3/graphql/develop/resolvers.md

```diff
@@ -42,7 +15,185 @@ $info | [`Magento\Framework\GraphQl\Schema\Type\ResolveInfo`]({{ site.mage2blobu
+
+Consider the following example:
+
+```graphql
+query ($filter: ProductAttributeFilterInput!) {
+  products (filter: $filter) {
+    items {
+      id
+      sku
+      related_products {
+        sku
+        related_products {
+          sku
+        }
+      }
+    }
+    total_count
+  }
+}
+```
+
+The query loads a list of products, the SKUs of their related products, and then any secondary related product SKUs.
+
+Loading a list of related products individually for each product would be expensive performance-wise. With batch resolvers, you can load linked products for all of products that were initially found, and then group them by root products. After the `items` branch is resolved, a batch resolver for `related_products` will be called for the first product found. But instead of resolving the list right away, it will just add the first product to the list of products that require loading additional related products. After all the products from the `items` branch have been loaded, the lists of related products must be loaded. Then `BatchResolverInterface::resolve()` executes with a gathered list of previous requests to `related_products` branches. At this point, the resolver is able to extract product DTOs from each GraphQL request, load all the product links, sort them by root products, and generate GraphQL values for each branch. After this is done, the same batching will take place when resolving child `related_products` branches.
+
+The following pseudo-code shows a related-products branch resolver:
+
+```php
+class RelatedProducts implements BatchResolverInterface
+{
+    ...
+
+    public function resolve(ContextInterface $context, Field $field, array $requests): BatchResponse
+    {
+        //Get the list of products we need to load related products for
+        $rootProductIds = array_map(function ($request) { return $request->getValue()['model']->getId(); }, $requests);
+
+        //Load the links
+        $productLinks = $this->service->getRelatedProductLinks($rootProductIds);
+
+        //Sort the links
+        $response = new BatchResponse();
+        foreach ($requests as $request) {
+            $response->addResponse($request, $productLinks[$request->getValue()['model']->getId()]);
+        }
+
+        return $response;
+    }
+}
+```
+
+Each GraphQL request object must be assigned a result of type [`\Magento\Framework\GraphQl\Query\Resolver\Value`]({{ site.mage2bloburl }}/{{ page.guide_version }}/lib/internal/Magento/Framework/GraphQl/Query/Resolver/Value.php) or any type of data (mixed). This value takes in a callable function to its constructor that will be invoked at the latest possible time for the resolver to require its data. As a result, a list of items being resolved can be retrieved all at once by establishing a buffer that contains all relevant parent data to filter and fetch for the children list data.
```

Comment:
```suggestion
Each GraphQL request object must be assigned a [`\Magento\Framework\GraphQl\Query\Resolver\Value`]({{ site.mage2bloburl }}/{{ page.guide_version }}/lib/internal/Magento/Framework/GraphQl/Query/Resolver/Value.php) result type or any type of data (mixed). This value takes in a callable function to its constructor that will be invoked at the latest possible time for the resolver to acquire its data. As a result, a list of items being resolved can be retrieved all at once by establishing a buffer that contains all relevant parent data to filter and fetch for the children list data.
```

Reply:
This paragraph was difficult for me to follow because there's so much passive voice in it. It might be more clear to someone familiar with the subject matter, but I had to re-read it several times.

Reply:
this bit already was in the doc, I just moved it